File: plantour-server/Services/PublicTemplatesService.cs
using plantour_server.DTOs;
using plantour_server.Repositories;
using plantour_server.Services.Interfaces;

namespace plantour_server.Services;

public class PublicTemplatesService(
    TemplateRepository templateRepository,
    AgeRangeRepository ageRangeRepository,
    TemperatureRangeRepository temperatureRangeRepository,
    ActivityRepository activityRepository) : IPublicTemplatesService
{
    private readonly TemplateRepository _templateRepository = templateRepository;
    private readonly AgeRangeRepository _ageRangeRepository = ageRangeRepository;
    private readonly TemperatureRangeRepository _temperatureRangeRepository = temperatureRangeRepository;
    private readonly ActivityRepository _activityRepository = activityRepository;

    public async Task<IEnumerable<PublicTemplateThingDto>> GetAllTemplateThingsAsync()
    {
        var entities = await _templateRepository.GetAllAsync();
        return entities.Select(MapTemplateThing);
    }

    public async Task<IEnumerable<PublicTemplateThingDto>> GetTemplateThingsByTemplateIdAsync(Guid templateId)
    {
        var entities = await _templateRepository.FindAsync(x => x.TemplateId == templateId);
        return entities.Select(MapTemplateThing);
    }

    public async Task<IEnumerable<PublicAgeRangeDto>> GetAgeRangesAsync()
    {
        var entities = await _ageRangeRepository.GetAllAsync();
        return entities.Select(x => new PublicAgeRangeDto
        {
            Id = x.Id,
            Name = x.Name,
            FromAge = x.Fromage,
            ToAge = x.Toage,
            Notes = x.Notes
        });
    }

    public async Task<IEnumerable<PublicTemperatureRangeDto>> GetTemperatureRangesAsync()
    {
        var entities = await _temperatureRangeRepository.GetAllAsync();
        return entities.Select(x => new PublicTemperatureRangeDto
        {
            Id = x.Id,
            Name = x.Name,
            FromTemp = x.Fromtemp,
            ToTemp = x.Totemp,
            Notes = x.Notes
        });
    }

    public async Task<IEnumerable<PublicActivityDto>> GetActivitiesAsync()
    {
        var entities = await _activityRepository.GetAllAsync();
        return entities.Select(x => new PublicActivityDto
        {
            Id = x.Id,
            Name = x.Name,
            Notes = x.Notes
        });
    }

    private static PublicTemplateThingDto MapTemplateThing(plantour_server.DbModels.VTemplateThingsFull entity)
    {
        return new PublicTemplateThingDto
        {
            ThingId = entity.ThingId ?? Guid.Empty,
            ThingName = entity.ThingName ?? string.Empty,
            Category = entity.Category,
            Units = entity.Units,
            Value = entity.Value,
            ThingNotes = entity.ThingNotes,
            TemplateId = entity.TemplateId ?? Guid.Empty,
            TemplateName = entity.TemplateName ?? string.Empty,
            ActivityName = entity.ActivityName ?? string.Empty,
            TemperatureRangeName = entity.TemperatureRangeName,
            FromTemp = entity.Fromtemp,
            ToTemp = entity.Totemp,
            AgeRangeName = entity.AgeRangeName,
            FromAge = entity.Fromage,
            ToAge = entity.Toage
        };
    }
}

File: plantour-client/src/app/services/public-templates-service.ts
import { HttpClient } from '@angular/common/http';
import { Inject, Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { ENVIRONMENT, EnvironmentConfig } from '../../environment.token';

export interface PublicTemplateThingDto {
  thingId: string;
  thingName: string;
  category?: string | null;
  units?: string | null;
  value?: number | null;
  thingNotes?: string | null;
  templateId: string;
  templateName: string;
  activityName: string;
  temperatureRangeName?: string | null;
  fromTemp?: number | null;
  toTemp?: number | null;
  ageRangeName?: string | null;
  fromAge?: number | null;
  toAge?: number | null;
}

export interface PublicAgeRangeDto {
  id: string;
  name: string;
  fromAge: number;
  toAge?: number | null;
  notes?: string | null;
}

export interface PublicTemperatureRangeDto {
  id: string;
  name: string;
  fromTemp?: number | null;
  toTemp?: number | null;
  notes?: string | null;
}

export interface PublicActivityDto {
  id: string;
  name: string;
  notes?: string | null;
}

@Injectable({
  providedIn: 'root'
})
export class PublicTemplatesService {
  private apiUrl: string;

  constructor(
    private http: HttpClient,
    @Inject(ENVIRONMENT) private environment: EnvironmentConfig
  ) {
    this.apiUrl = `${environment.api.baseUrl}/public`;
  }

  getTemplateThings(): Observable<PublicTemplateThingDto[]> {
    return this.http.get<PublicTemplateThingDto[]>(`${this.apiUrl}/templates`);
  }

  getTemplateThingsByTemplateId(templateId: string): Observable<PublicTemplateThingDto[]> {
    return this.http.get<PublicTemplateThingDto[]>(`${this.apiUrl}/templates/${templateId}`);
  }

  getAgeRanges(): Observable<PublicAgeRangeDto[]> {
    return this.http.get<PublicAgeRangeDto[]>(`${this.apiUrl}/age-ranges`);
  }

  getTemperatureRanges(): Observable<PublicTemperatureRangeDto[]> {
    return this.http.get<PublicTemperatureRangeDto[]>(`${this.apiUrl}/temperature-ranges`);
  }

  getActivities(): Observable<PublicActivityDto[]> {
    return this.http.get<PublicActivityDto[]>(`${this.apiUrl}/activities`);
  }
}
